akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala

```diff
@@ -169,49 +173,44 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
 
   private def expectNoMessage_internal(max: FiniteDuration): Unit = {
     val o = receiveOne(max)
-    assert(o == null, s"received unexpected message $o")
-    lastWasNoMessage = true
+    o match {
+      case None    ⇒ lastWasNoMessage = true
+      case Some(o) ⇒ assertFail(s"received unexpected message $o")
+    }
   }
 
   override protected def expectMessageClass_internal[C](max: FiniteDuration, c: Class[C]): C = {
     val o = receiveOne(max)
-    assert(o != null, s"timeout ($max) during expectMessageClass waiting for $c")
-    assert(BoxedType(c) isInstance o, s"expected $c, found ${o.getClass} ($o)")
-    o.asInstanceOf[C]
+    val bt = BoxedType(c)
+    o match {
+      case Some(o) if bt isInstance o ⇒ o.asInstanceOf[C]
+      case Some(o)                    ⇒ assertFail(s"expected $c, found ${o.getClass} ($o)")
+      case None ⇒ assertFail(
+        s"timeout ($max) during expectMessageClass waiting for $c")
+    }
   }
 
   override protected def fishForMessage_internal(max: FiniteDuration, hint: String, fisher: M ⇒ FishingOutcome): List[M] = {
-    // not tailrec but that should be ok
-    def loop(timeout: FiniteDuration, seen: List[M]): List[M] = {
+    @tailrec def loop(timeout: FiniteDuration, seen: List[M]): List[M] = {
       val start = System.nanoTime()
-      val maybeMsg = Option(receiveOne(timeout))
+      val maybeMsg = receiveOne(timeout)
       maybeMsg match {
         case Some(message) ⇒
-          try {
-            fisher(message) match {
-              case FishingOutcome.Complete    ⇒ (message :: seen).reverse
-              case FishingOutcome.Fail(error) ⇒ throw new AssertionError(s"$error, hint: $hint")
-              case continue ⇒
-                val newTimeout =
-                  if (timeout.isFinite()) timeout - (System.nanoTime() - start).nanos
-                  else timeout
-                if (newTimeout.toMillis <= 0) {
-                  throw new AssertionError(s"timeout ($max) during fishForMessage, seen messages ${seen.reverse}, hint: $hint")
-                } else {
-
-                  continue match {
-                    case FishingOutcome.Continue          ⇒ loop(newTimeout, message :: seen)
-                    case FishingOutcome.ContinueAndIgnore ⇒ loop(newTimeout, seen)
-                    case _                                ⇒ ??? // cannot happen
-                  }
-
-                }
-            }
-          } catch {
+          val outcome = try fisher(message) catch {
             case ex: MatchError ⇒ throw new AssertionError(
               s"Unexpected message $message while fishing for messages, " +
                 s"seen messages ${seen.reverse}, hint: $hint", ex)
           }
+          outcome match {
+            case FishingOutcome.Complete    ⇒ (message :: seen).reverse
+            case FishingOutcome.Fail(error) ⇒ throw new AssertionError(s"$error, hint: $hint")
```

Comment:
is this intended to not use `assertFail`?

Reply:
Yes because of the leading "assertion failed: " that gets added with the old `assert` conditions.  But that's not so important, so I'll reuse `assertFail` more.